src/jdbc/cubrid/jdbc/jci/UTimedDataInputStream.java

```diff
@@ -49,101 +49,118 @@
+        long begin = System.currentTimeMillis();
+
+        while (true) {
+          try {
+            stream.readFully(b);
+            return;
+          } catch (SocketTimeoutException e) {
+            if (timeout > 0 && timeout - (System.currentTimeMillis() - begin) <= 0) {
+              String msg = UErrorCode.codeToMessage(UErrorCode.ER_TIMEOUT);
+              throw new SocketTimeoutException(msg);
+            }
+            BrokerHandler.pingBroker(ip, port, PING_TIMEOUT);
+          }
+        }
     }
 
-    public int readByte(byte[] b, int timeout) throws IOException {
-	long begin = System.currentTimeMillis();
-
-	while (true) {
-	    try {
-		return stream.read(b);
-	    } catch (SocketTimeoutException e) {
-		if (timeout > 0 && timeout - (System.currentTimeMillis() - begin) <= 0) {
-		    String msg = UErrorCode.codeToMessage(UErrorCode.ER_TIMEOUT);
-		    throw new SocketTimeoutException(msg);
-		}
-		BrokerHandler.pingBroker(ip, port, PING_TIMEOUT);
-	    }
-	}
+    public int readByte(byte[] b, int timeout) throws IOException, UJciException {
+        long begin = System.currentTimeMillis();
+
+        while (true) {
+          try {
+            return stream.read(b);
+          } catch (SocketTimeoutException e) {
+            if (timeout > 0 && timeout - (System.currentTimeMillis() - begin) <= 0) {
+              String msg = UErrorCode.codeToMessage(UErrorCode.ER_TIMEOUT);
+              throw new SocketTimeoutException(msg);
+            }
+            BrokerHandler.pingBroker(ip, port, PING_TIMEOUT);
+          }
+        }
     }
 
-    public int readByte(byte[] b) throws IOException {
-	return readByte(b, timeout);
+    public int readByte(byte[] b) throws IOException, UJciException {
+        return readByte(b, timeout);
     }
 
-    public int read(byte[] b, int off, int len, int timeout) throws IOException {
-	long begin = System.currentTimeMillis();
-
-	while (true) {
-	    try {
-		return stream.read(b, off, len);
-	    } catch (SocketTimeoutException e) {
-		if (timeout > 0 && timeout - (System.currentTimeMillis() - begin) <= 0) {
-		    String msg = UErrorCode.codeToMessage(UErrorCode.ER_TIMEOUT);
-		    throw new SocketTimeoutException(msg);
-		}
-		BrokerHandler.pingBroker(ip, port, PING_TIMEOUT);
-	    }
-	}
+    public int read(byte[] b, int off, int len, int timeout) throws IOException, UJciException {
+        long begin = System.currentTimeMillis();
+        boolean retry = false;
+
+        while (true) {
+          try {
+            return stream.read(b, off, len);
+          } catch (SocketTimeoutException e) {
+            if (timeout > 0 && timeout - (System.currentTimeMillis() - begin) <= 0) {
+              String msg = UErrorCode.codeToMessage(UErrorCode.ER_TIMEOUT);
+              throw new SocketTimeoutException(msg);
+            }
+            if (BrokerHandler.statusBroker(ip, port, pid, session, PING_TIMEOUT) != 1) {
```

Comment:
How about letting statusBroker perform by adding JDBC properties like 'casHealthChk'? default is pingBroker.
Here's why I think so:
1. Endless wait is not a common problem.
2. brokerStatus is heavier than brokerPing.
3. There may be unexpected side effects.

Reply:
I will refer your option in the next release but we need to release patch version first.
So, we need to check the daily QA result.